laws/src/main/scala/cats/laws/FunctorLaws.scala

```diff
@@ -33,6 +33,12 @@ trait FunctorLaws[F[_]] extends InvariantLaws[F] {
   def covariantIdentity[A](fa: F[A]): IsEq[F[A]] =
     fa.map(identity) <-> fa
 
+  def mapOrKeepEmpty[A](fa: F[A]): IsEq[F[A]] =
+    fa.mapOrKeep(PartialFunction.empty) <-> fa
+
+  def mapOrKeep[A](fa: F[A], f: A => A): IsEq[F[A]] =
+    fa.mapOrKeep { case a => f(a) } <-> fa.map(f)
```

Comment:
@jozic , @johnynek would it make sense to add a more generic rule, like:
```scala
def mapOrKeepToMapEquivalence[A, A1 >: A](fa: F[A], pf: PartialFunction[A, A1]): IsEq[F[A1]] =
  Functor[A].mapOrKeep(fa)(pf) <-> Functor[A].map(fa)(a => pf.applyOrElse(a, _ => a: A1)
```
or something like that?

Reply:
@satorg 
i didn't find any examples in the repo of how to generate `A` and `A1` so that `A1 >: A`
do you have an idea how to do that?

Reply:
@satorg 
i can add 
```scala
  def mapOrKeepToMapEquivalence[A, A1 >: A](fa: F[A], pf: PartialFunction[A, A1]): IsEq[F[A1]] =
    fa.mapOrKeep(pf) <-> fa.map(a => pf.applyOrElse(a, identity[A1]))
```
 and call it with 
```scala
      "mapOrKeepToMapEquivalence" -> forAll(laws.mapOrKeepToMapEquivalence[A, A] _)
```
will that be good enough?

Reply:
Yes, I guess it would be. But I've just realized, there is a couple of concerns about that:
1. If we get the `mapOrKeepToMapEquivalence` law, do we really need the other two? To me the `mapOrKeepToMapEquivalence` one seems implying the former two. I feel since this method is a derivative from `map` then it should not requite that many laws.
2. I am not sure that it will be easy to add the `A1 >: A` type to the `FunctorTests.functor` rule set – simple addition of one more parameter with all necessary implicits can make Mima unhappy. In that case it would be reasonable to go with `PartialFunction[A, A]` there as well.

Furthermore, if we'd agree to add the `mapOrKeepToMapEquivalence` law, then it might not make a lot of sense to have the other test we're discussing – `mapOrKeep maps matching elements preserving structure` in `FunctorSuite`. Apparently, the latter would be nothing but a duplication of the law. So I believe we could get rid of the latter test completely – the `Functor` laws are already executed for many different types including `List`, `Map`, etc.

Reply:
added `mapOrKeepToMapEquivalence` and removed other laws and tests in `FunctorSuite`